src/com/goide/formatter/GoOnSaveFormatter.java

```diff
@@ -0,0 +1,177 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.goide.formatter;
+
+import com.goide.GoEnvironmentUtil;
+import com.goide.GoFileType;
+import com.goide.sdk.GoSdkUtil;
+import com.goide.settings.GoProjectSettings;
+import com.goide.ui.GoToolWindow;
+import com.intellij.execution.process.OSProcessHandler;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.fileEditor.FileDocumentManager;
+import com.intellij.openapi.fileEditor.FileDocumentManagerAdapter;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.projectRoots.Sdk;
+import com.intellij.openapi.roots.ProjectRootManager;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.openapi.util.SystemInfo;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.openapi.vfs.VirtualFileManager;
+import com.intellij.util.EnvironmentUtil;
+import org.jetbrains.annotations.NotNull;
+
+
+public class GoOnSaveFormatter extends FileDocumentManagerAdapter {
+  @Override
+  public void beforeDocumentSaving(@NotNull final Document document) {
+
+    if (!document.isWritable()) {
+      return;
+    }
+
+    final Project[] projects = ProjectManager.getInstance().getOpenProjects();
+    if (projects.length == 0) {
+      return;
+    }
+
+    final VirtualFile file = FileDocumentManager.getInstance().getFile(document);
+    if (file == null || file.getFileType() != GoFileType.INSTANCE) {
+      return;
+    }
+
+    Project project = null;
+    for (Project possibleProject : projects) {
+      if (ProjectRootManager.getInstance(possibleProject).getFileIndex().getSourceRootForFile(file) != null ||
+          ProjectRootManager.getInstance(possibleProject).getFileIndex().getContentRootForFile(file) != null) {
+        project = possibleProject;
+        break;
+      }
+    }
+
+    if (project == null) {
+      return;
+    }
+
+    GoProjectSettings.GoProjectSettingsBean settings = GoProjectSettings.getInstance(project).getState();
+
+    final Project p = project;
+    if (settings.goimportsOnSave) {
+      ApplicationManager.getApplication().invokeLater(new Runnable() {
+        @Override
+        public void run() {
+          if (p.isDisposed()) {
+            return;
+          }
+          ProcessFileWithGoImports(p, file);
+        }
+      });
+    }
+    else if (settings.goFmtOnSave) {
+      ApplicationManager.getApplication().invokeLater(new Runnable() {
+        @Override
+        public void run() {
+          if (p.isDisposed()) {
+            return;
+          }
+          ProcessFileWithGoFmt(p, file);
```

Comment:
We can reuse com.goide.actions.fmt.GoFmtFileAction and com.goide.actions.fmt.GoImportsFileAction